File: Examples/Examples/API/APIClient.m
//
//  APIClient.m
//  Examples
//
//  Created by Victor Zhu on 2020/1/15.
//  Copyright © 2020 Airwallex. All rights reserved.
//

#import "APIClient.h"
#import <Airwallex/Airwallex.h>

@implementation APIClient

+ (instancetype)sharedClient
{
    static APIClient *sharedClient;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedClient = [self new];
    });
    return sharedClient;
}

- (void)setPaymentBaseURL:(NSURL *)paymentBaseURL
{
    _paymentBaseURL = [paymentBaseURL URLByAppendingPathComponent:@""];
}

- (void)createAuthenticationTokenWithCompletionHandler:(void (^)(NSError * _Nullable error))completionHandler
{
    NSURL *requestURL = [NSURL URLWithString:@"api/v1/authentication/login" relativeToURL:self.paymentBaseURL];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:requestURL];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request setValue:self.clientID forHTTPHeaderField:@"x-client-id"];
    [request setValue:self.apiKey forHTTPHeaderField:@"x-api-key"];
    [[[NSURLSession sharedSession] dataTaskWithRequest:request
                                     completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(error);
            });
            return;
        }
        
        NSError *anError;
        if (data) {
            NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data
                                                                 options:NSJSONReadingMutableContainers
                                                                   error:&anError];
            NSString *errorMessage = json[@"error"];
            if (errorMessage) {
                anError = [NSError errorWithDomain:@"com.airwallex.paymentacceptance" code:-1 userInfo:@{NSLocalizedDescriptionKey: errorMessage}];
            }
            self.token = json[@"token"];
            
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(anError);
            });
            return;
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler(anError);
        });
    }] resume];
}

- (void)createPaymentIntentWithParameters:(NSDictionary *)parameters
                        completionHandler:(void (^)(AWXPaymentIntent * _Nullable paymentIntent, NSError * _Nullable error))completionHandler
{
    NSURL *requestURL = [NSURL URLWithString:@"api/v1/pa/payment_intents/create" relativeToURL:self.paymentBaseURL];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:requestURL];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    if (self.token) {
        [request setValue:[NSString stringWithFormat:@"Bearer %@", self.token] forHTTPHeaderField:@"Authorization"];
    }
    [request setHTTPBody:[NSJSONSerialization dataWithJSONObject:parameters options:NSJSONWritingPrettyPrinted error:nil]];
    [[[NSURLSession sharedSession] dataTaskWithRequest:request
                                     completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(nil, error);
            });
            return;
        }
        
        NSError *anError;
        if (data) {
            NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data
                                                                 options:NSJSONReadingMutableContainers
                                                                   error:&anError];
            NSString *errorMessage = json[@"message"];
            if (errorMessage) {
                anError = [NSError errorWithDomain:@"com.airwallex.paymentacceptance" code:-1 userInfo:@{NSLocalizedDescriptionKey: errorMessage}];
            }
            
            AWXPaymentIntent *paymentIntent = [AWXPaymentIntent decodeFromJSON:json];
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(paymentIntent, anError);
            });
            return;
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler(nil, anError);
        });
    }] resume];
}

- (void)createCustomerWithParameters:(NSDictionary *)parameters
                   completionHandler:(void (^)(NSDictionary * _Nullable result, NSError * _Nullable error))completionHandler
{
    NSURL *requestURL = [NSURL URLWithString:@"api/v1/pa/customers/create" relativeToURL:self.paymentBaseURL];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:requestURL];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    if (self.token) {
        [request setValue:[NSString stringWithFormat:@"Bearer %@", self.token] forHTTPHeaderField:@"Authorization"];
    }
    [request setHTTPBody:[NSJSONSerialization dataWithJSONObject:parameters options:NSJSONWritingPrettyPrinted error:nil]];
    [[[NSURLSession sharedSession] dataTaskWithRequest:request
                                     completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(nil, error);
            });
            return;
        }
        
        NSError *anError;
        if (data) {
            NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data
                                                                 options:NSJSONReadingMutableContainers
                                                                   error:&anError];
            NSString *errorMessage = json[@"message"];
            if (errorMessage) {
                anError = [NSError errorWithDomain:@"com.airwallex.paymentacceptance" code:-1 userInfo:@{NSLocalizedDescriptionKey: errorMessage}];
            }
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(json, anError);
            });
            return;
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler(nil, anError);
        });
    }] resume];
}

- (void)createCustomerSecretWithId:(NSString *)Id
                 completionHandler:(void (^)(NSDictionary * _Nullable result, NSError * _Nullable error))completionHandler
{
    NSURL *requestURL = [NSURL URLWithString:[NSString stringWithFormat:@"api/v1/pa/customers/%@/generate_client_secret", Id] relativeToURL:self.paymentBaseURL];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:requestURL];
    [request setHTTPMethod:@"GET"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    if (self.token) {
        [request setValue:[NSString stringWithFormat:@"Bearer %@", self.token] forHTTPHeaderField:@"Authorization"];
    }
    [[[NSURLSession sharedSession] dataTaskWithRequest:request
                                     completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(nil, error);
            });
            return;
        }
        
        NSError *anError;
        if (data) {
            NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data
                                                                 options:NSJSONReadingMutableContainers
                                                                   error:&anError];
            NSString *errorMessage = json[@"message"];
            if (errorMessage) {
                anError = [NSError errorWithDomain:@"com.airwallex.paymentacceptance" code:-1 userInfo:@{NSLocalizedDescriptionKey: errorMessage}];
            }
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(json, anError);
            });
            return;
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            completionHandler(nil, anError);
        });
    }] resume];
}

@end

File: Examples/Examples/API/APIClient.swift
//
//  APIClient.swift
//  Examples
//
//  Created by Hector.Huang on 2024/9/5.
//  Copyright © 2024 Airwallex. All rights reserved.
//

import Foundation
import Airwallex

protocol APIClient {
    func createPaymentIntent(request: PaymentIntentRequest, completion: @escaping (Result<AWXPaymentIntent, Error>) -> Void)
    
    func generateClientSecret(customerID: String, apiKey: String?, clientID: String?, completion: @escaping (Result<String, Error>) -> Void)
    
    func createCustomer(request: CustomerRequest, completion: @escaping (Result<Customer, Error>) -> Void)
}

extension APIClient {
    
    func createPaymentIntent(force3DS: Bool = false) async throws -> AWXPaymentIntent {
        let request = PaymentIntentRequest(
            amount: Decimal(string: ExamplesKeys.amount) ?? 0,
            currency: ExamplesKeys.currency,
            order: DemoDataSource.createOrder(),
            metadata: ["id": 1],
            returnUrl: ExamplesKeys.returnUrl,
            customerID: ExamplesKeys.customerId?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty,
            paymentMethodOptions: force3DS ? ["card": ["three_ds_action": "FORCE_3DS"]] : nil,
            apiKey: ExamplesKeys.apiKey,
            clientID: ExamplesKeys.clientId
        )
        
        let paymentIntent = try await withCheckedThrowingContinuation { continuation in
            createPaymentIntent(request: request) { continuation.resume(with: $0) }
        }
        return paymentIntent
    }
}

extension Airwallex {
    static var apiClient: APIClient {
        switch Airwallex.mode() {
        case .demoMode, .stagingMode:
            return DemoStoreAPIClient()
        case .productionMode:
            return DirectAPIClient()
        }
    }
}

@objc protocol CustomerFetchable {
    func createCustomer(
        firstName: String?,
        lastName: String?,
        email: String?,
        phoneNumber: String?,
        additionalInfo: Dictionary<String, Any>?,
        metadata: Dictionary<String, Int>,
        apiKey: String?,
        clientID: String?,
        completion: @escaping (Customer?, Error?) -> Void
    )
}

/// For demo/staging environment if clientId/apiKey is not set, airwallex demo/staging server will use default value configured on server
/// If you configure it in on client side it will override default configuration on server
///
/// `DirectAPIClient` is for debugging purpose in production environment.
///  For your app, you are recommended to config clientId/apiKey on your own server and let your server
///  and let your server do the token creation and other communitation with airwallex service for your app
class DemoStoreAPIClient: APIClient, CustomerFetchable {
    
    init(baseURL: String? = nil) {
        demoStoreBaseUrl = baseURL ?? Self.baseURLForEnvironment(Airwallex.mode())
    }
    
    static func baseURLForEnvironment(_ environment: AirwallexSDKMode) -> String? {
        switch environment {
        case .demoMode:
            "https://demo-pacheckoutdemo.airwallex.com"
        case .stagingMode:
            "https://staging-pacheckoutdemo.airwallex.com"
        case .productionMode:
            // Our demo does not support production. Please validate it by connecting your own backend.
            nil
        }
    }
    
    private var demoStoreBaseUrl: String?
    
    func createPaymentIntent(request: PaymentIntentRequest, completion: @escaping (Result<AWXPaymentIntent, Error>) -> Void) {
        post(path: "/api/v1/pa/payment_intents/create", encodable: request, completion: completion)
    }
    
    func createCustomer(request: CustomerRequest, completion: @escaping (Result<Customer, Error>) -> Void) {
        post(path: "/api/v1/pa/customers/create", encodable: request, completion: completion)
    }
    
    func createCustomer(firstName: String?, lastName: String?, email: String?, phoneNumber: String?, additionalInfo: Dictionary<String, Any>?, metadata: Dictionary<String, Int>, apiKey: String?, clientID: String?, completion: @escaping (Customer?, Error?) -> Void) {
        createCustomer(request: CustomerRequest(firstName: firstName, lastName: lastName, email: email, phoneNumber: phoneNumber, additionalInfo: additionalInfo, metadata: metadata, apiKey: apiKey, clientID: clientID)) { result in
            switch (result) {
            case .success(let customer):
                completion(customer, nil)
            case .failure(let error):
                completion(nil, error)
            }
        }
    }
    
    func generateClientSecret(customerID: String, apiKey: String?, clientID: String?, completion: @escaping (Result<String, any Error>) -> Void) {
        guard let baseUrl = demoStoreBaseUrl, var urlComponents = URLComponents(string: baseUrl + "/api/v1/pa/customers/\(customerID)/generate_client_secret") else {
            preconditionFailure("Unable to build URL components")
        }
        var queryItems = [URLQueryItem]()
        if let apiKey {
            queryItems.append(.init(name: "apiKey", value: apiKey))
        }
        if let clientID {
            queryItems.append(.init(name: "clientId", value: clientID))
        }
        urlComponents.queryItems = queryItems
        
        guard let url = urlComponents.url else {
            preconditionFailure("Unable to unwrap URL")
        }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        URLSession.shared.dataTask(with: request) { data, _, error in
            if let error {
                completion(.failure(error))
                return
            }
            if let data {
                let responseDict: [String: Any]?
                do {
                    responseDict = try JSONSerialization.jsonObject(with: data) as? [String: Any]
                } catch {
                    completion(.failure(error))
                    return
                }
                
                if let errorMessage = responseDict?["message"] as? String {
                    completion(.failure(NSError.airwallexError(localizedMessage: errorMessage)))
                    return
                }
                if let clientSecret = (responseDict as? [String: String])?["client_secret"] {
                    completion(.success(clientSecret))
                } else {
                    completion(.failure(NSError.airwallexError(localizedMessage: "Failed to decode to client secret")))
                }
            }
        }.resume()
    }
    
    private func post<T: AWXJSONDecodable>(path: String, encodable: Encodable, completion: @escaping (Result<T, Error>) -> Void) {
        guard let baseUrl = demoStoreBaseUrl, let url = URL(string: baseUrl + path) else {
            preconditionFailure("Unable to unwrap URL")
        }
        guard let body = try? JSONEncoder().encode(encodable) else {
            preconditionFailure("Unable to serialize JSON")
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        URLSession.shared.dataTask(with: request) { data, _, error in
            if let error {
                completion(.failure(error))
                return
            }
            if let data {
                let responseDict: [String: Any]?
                do {
                    responseDict = try JSONSerialization.jsonObject(with: data) as? [String: Any]
                } catch {
                    completion(.failure(error))
                    return
                }
                
                if let errorMessage = responseDict?["message"] as? String {
                    completion(.failure(NSError.airwallexError(localizedMessage: errorMessage)))
                } else {
                    if let model = T.decode(fromJSON: responseDict) as? T {
                        completion(.success(model))
                    } else {
                        completion(.failure(NSError.airwallexError(localizedMessage: "Failed to decode to \(String(describing: T.self))")))
                    }
                }
            }
        }.resume()
    }
}
